GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs

```diff
@@ -21,27 +24,72 @@ public LooseObjectsStep(GVFSContext context, bool requireCacheLock, bool forceRu
         }
 
         public override string Area => nameof(LooseObjectsStep);
+        public int LooseObjectsPutIntoPackFile { get; private set; } = 0;
+        public int MaxLooseObjectsInPack { get; set; } = 100000;
 
         protected override string LastRunTimeFilePath => Path.Combine(this.Context.Enlistment.GitObjectsRoot, "info", LooseObjectsLastRunFileName);
         protected override TimeSpan TimeBetweenRuns => TimeSpan.FromDays(7);
 
         public int CountLooseObjects()
         {
             int count = 0;
-            
+
             foreach (string directoryPath in this.Context.FileSystem.EnumerateDirectories(this.Context.Enlistment.GitObjectsRoot))
             {
                 string directoryName = directoryPath.TrimEnd(Path.DirectorySeparatorChar).Split(Path.DirectorySeparatorChar).Last();
 
                 if (GitObjects.IsLooseObjectsDirectory(directoryName))
-                {                    
+                {
                     count += this.Context.FileSystem.GetFiles(Path.Combine(this.Context.Enlistment.GitObjectsRoot, directoryPath), "*").Count();
                 }
             }
 
             return count;
         }
 
+        public void PackLooseObjects(StreamWriter streamWriter)
+        {
+            // Find loose Objects
+            foreach (DirectoryItemInfo directoryItemInfo in this.Context.FileSystem.ItemsInDirectory(this.Context.Enlistment.GitObjectsRoot))
+            {
+                if (directoryItemInfo.IsDirectory)
+                {
+                    string directoryName = directoryItemInfo.Name;
+
+                    if (GitObjects.IsLooseObjectsDirectory(directoryName))
+                    {
+                        string[] looseObjectFileNamesInDir = this.Context.FileSystem.GetFiles(Path.Combine(directoryItemInfo.FullName), "*");
+
+                        foreach (string file in looseObjectFileNamesInDir)
+                        {
+                            this.LooseObjectsPutIntoPackFile++;
+                            streamWriter.Write(this.GetLooseObjectId(directoryName, file) + "\n");
+
+                            // Stop when we hit the limit.  The next run will pack more files.
+                            if (this.LooseObjectsPutIntoPackFile == this.MaxLooseObjectsInPack)
+                            {
+                                return;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        public string GetLooseObjectId(string directoryName, string filePath)
+        {
+            return directoryName + Path.GetFileName(filePath);
+        }
+
+        public void CreateLooseObjectsPackFile()
+        {
+            GitProcess.Result result = this.RunGitCommand(
+                (process) => process.PackObjects(
+                    "from-loose",
+                    this.Context.Enlistment.GitObjectsRoot,
+                    this.PackLooseObjects));

```

Comment:
What will happen if `PackLooseObjects` finds no loose objects?

Reply:
PackLooseObjects calls  $"pack-objects {packFilePath} --non-empty -q".  This means if there are no loose objects don't create a pack.  This case has been tested.  Would be a good idea to add a function test for it.  I'll make it happen!